crates/iroha_test_network/src/lib.rs

```diff
@@ -1,883 +1,792 @@
 
-        (peer, client)
+    /// Build and start the network.
+    ///
+    /// Resolves when all peers are running and have committed genesis block.
+    /// See [`Network::start_all`].
+    pub async fn start(self) -> Result<Network> {
+        let network = self.build();
+        network.start_all().await;
+        Ok(network)
     }
 
-    /// Create a peer with a client, create a runtime, and synchronously start the peer on the runtime.
-    pub fn start_with_runtime(self) -> PeerWithRuntimeAndClient {
-        let rt = Runtime::test();
-        let (peer, client) = rt.block_on(self.start_with_client());
-        (rt, peer, client)
+    /// Combination of [`Self::build_blocking`] and [`Self::start`].
+    pub fn start_blocking(self) -> Result<(Network, Runtime)> {
+        let (network, rt) = self.build_blocking();
+        rt.block_on(async { network.start_all().await });
+        Ok((network, rt))
     }
 }
 
-type PeerWithRuntimeAndClient = (Runtime, Peer, Client);
+/// A common signatory in the test network.
+///
+/// # Example
+///
+/// ```
+/// use iroha_test_network::Signatory;
+///
+/// let _alice_kp = Signatory::Alice.key_pair();
+/// ```
+pub enum Signatory {
+    Peer,
+    Genesis,
+    Alice,
+}
 
-fn local_unique_port() -> Result<SocketAddr> {
-    Ok(socket_addr!(127.0.0.1: unique_port::get_unique_free_port().map_err(eyre::Error::msg)?))
+impl Signatory {
+    /// Get the associated key pair
+    pub fn key_pair(&self) -> &KeyPair {
+        match self {
+            Signatory::Peer => &PEER_KEYPAIR,
+            Signatory::Genesis => &SAMPLE_GENESIS_ACCOUNT_KEYPAIR,
+            Signatory::Alice => &ALICE_KEYPAIR,
+        }
+        .deref()
+    }
 }
 
-/// Runtime used for testing.
-pub trait TestRuntime {
-    /// Create test runtime
-    fn test() -> Self;
+/// Running Iroha peer.
+///
+/// Aborts peer forcefully when dropped
+struct PeerRun {
+    tasks: JoinSet<()>,
+    kill: oneshot::Sender<()>,
 }
 
-/// Peer configuration mocking trait.
-pub trait TestConfig {
-    /// Creates test configuration
-    fn test() -> Self;
-    /// Returns default pipeline time.
-    fn pipeline_time() -> Duration;
-    /// Returns default time between block sync requests
-    fn block_sync_gossip_time() -> Duration;
+#[derive(Display, Clone)]
+#[display(fmt = "[p2p: {p2p}, api: {api}, dir: {dir}, run: #{run}]")]
+struct RunId {
+    p2p: u16,
+    api: u16,
+    dir: String,
+    run: usize,
 }
 
-/// Client configuration mocking trait.
-pub trait TestClientConfig {
-    /// Creates test client configuration
-    fn test(api_address: &SocketAddr) -> Self;
+/// Lifecycle events of a peer
+#[derive(Copy, Clone, Debug)]
+pub enum PeerLifecycleEvent {
+    /// Process spawned
+    Spawned,
+    /// Server started to respond
+    ServerStarted,
+    /// Process terminated
+    Terminated { status: ExitStatus },
+    /// Process was killed
+    Killed,
+    /// Caught a "Block committed" log message
+    LogBlockCommitted { height: u64 },
```

Comment:
That's a good point for discussion.

In my understanding, logging could be generally split on two categories:

1. Part of the public interface. This makes sense to test, as users might rely on it. Usually indicates some core domain events. 
2. Debugging. No need to test, should not be relied upon, could always be changed. 

I thought that the "Block committed" message is quite an important one in logs and it makes sense to rely on it in tests. If we change it in future, well, (1) it isn't hard to update tests and (2) thanks to tests we will see that something important in Iroha UX changed. 

Another example of an important message that should be, imo, standardised and fixed, is an error when the genesis is invalid. Currently, Iroha just panics and exits with messy logs. Instead, it should be a proper error message like "Genesis is invalid. This is a fatal error, terminating Iroha. Try to fix it by this and that.". That, in turn, could be a clear lifecycle hook here, in `iroha_test_network`. And it could be used by the `genesis_transactions_are_validated_by_executor`, which will wait for a precise failure reason.

Reply:
if we can avoid relying on log we should do it. Can we have an event listener that listens for `BlockCommitted` event in this case?

Reply:
Yeah, using an event listener would be more reliable in this case.

Reply:
Used events listener instead.